### DISCUSSION TITLE
Remove covariance on Arrays

### DISCUSSION THREAD

**Discussion by hez2010** [dotnet/csharplang#5153] [Language Ideas]
## Remove covariance on Arrays

Considerations: 

- Convariance on Arrays is preventing lots of JIT optimizations
- Convariance on Arrays is useless, and almost no one is actually using it
- Convariance on Arrays is breaking type system

```csharp

Base[] array = new Derived[10];
array[0] = new Base(); // exception here, type system is broken

class Base {}
class Derived : Base {}
```

**Comment by CyrusNajmabadi** [comment]
> Covariance on Arrays is useless, and almost no one is actually using it

We would need evidence of this.  Absent that, this is going to be too big a breaking change to be realistically taken.

**Reply by hez2010** [reply]
How about introducing a feature flag to disable covariance on arrays only when the flag is `true`? Then we can see how removing covariance on arrays is affecting users.

**Reply by huoyaoyuan** [reply]
The JIT should be able to mark all array variance conversions. We can start from identifying them.

**Comment by alrz** [comment]
I think if the concept of a "readonly array" is exposed by language/runtime, this could be done only for those?

**Reply by svick** [reply]
Is that actually needed, when we already have `ReadOnlySpan<T>` and `ReadOnlyMemory<T>`? They aren't covariant, so they don't suffer the same performance issues.

**Reply by alrz** [reply]
So is `ImmutableArray<T>` with an underlying `T[]` but I don't think the runtime knows about that.

**Reply by huoyaoyuan** [reply]
Span is slightly less efficient than array since it carries two fields.
ImmutableArray is efficient, but it uses array covariance in casting.

**Reply by gfoidl** [reply]
> Span is slightly less efficient than array since it carries two fields

Span is intrinsic to the JIT, so if the JIT does a good job it shouldn't be less efficient that array, rather be on par.

**Comment by hez2010** [comment]
An interesting case:

```csharp
Random[] a = new Random[1];
object[] c = a;
MemoryMarshal.GetArrayDataReference(c) = "a string";
var l = a[0].Next(); // guess what method is actually being called
Console.WriteLine(l);
```
Array covariance makes code completely unsafe and can lead to undefined behavior, and its benefit-harm ratio is minus: 
Benefits:
- You can upcast an array

Harm:
- It can easily bypass the compile-time type check without a line of unsafe code
- It breaks the type system and leads to runtime exception of type mismatch despite successful compilation
- It introduces unnecessary runtime check against type of elements in an array which introduces more runtime overhead
- It makes JIT hard to optimize against arrays
- It can lead to undefined behavior easily and can bypass the runtime check

**Reply by HaloFour** [reply]
I think most people would agree that copying covariance of arrays from Java was a bad idea and that if .NET were to be redesigned today that this would not be allowed.  But 20 years in that's a massive potential breaking change, one that affects the runtime as well as every .NET language.  At this point an analyzer which strongly recommends not trying to upcast arrays is probably a better plan.

**Reply by hez2010** [reply]
An analyzer can help prevent users writing array upcast, but the runtime will still need to proceed those unnecessary checks which leads to more overhead. I think if we take the analyzer approach, there should be an option available to turn off those checks in the runtime.

Otherwise, we have to cast arrays to `Span<T>` for writing to achieve higher performance. You can see huge performance difference between `Span<T>` and arrays, and you can see the improvement only with a `AsSpan()`. So it raises an embarrassing situation: if everyone need to use `AsSpan()` for array writes, why array is providing its own indexer for write access?

**Reply by HaloFour** [reply]
I don't think that option is practical.  The removal of the runtime checks without assurances only leads to further unsafe code that could have very subtle problems.

From what I've read of [this conversation](https://github.com/dotnet/runtime/issues/41565), if you're writing performance-heavy code where array covariance is of concern that the recommendation is to use `Span<T>` anyway.  A [separate effort](https://github.com/dotnet/coreclr/pull/23571) to come up with a separate invariant array type also lost steam.

**Comment by huoyaoyuan** [comment]
Some interesting benchmark I've done:
```csharp
    class Base
    {
    }

    class Derived : Base
    {
    }

    [SimpleJob(RuntimeMoniker.Net60)]
    public class Program
    {
        static void Main(string[] args)
        {
            BenchmarkRunner.Run<Program>();
        }

        private readonly Base[] array = new Base[256];
        private Derived derived = new();
        private Base @base = new();
        private readonly int[] intArray = new int[256];

        [Benchmark]
        public object Array()
        {
            for (int i = 0; i < array.Length; i++)
                array[i] = derived;
            return array;
        }

        [Benchmark]
        public object ArrayBase()
        {
            for (int i = 0; i < array.Length; i++)
                array[i] = @base;
            return array;
        }
        
        [Benchmark]
        public object ArrayInt()
        {
            for (int i = 0; i < intArray.Length; i++)
                intArray[i] = 19260817;
            return intArray;
        }

        [Benchmark]
        public object SpanIndex()
        {
            Span<Base> span = array;
            for (int i = 0; i < span.Length; i++)
                span[i] = derived;
            return array;
        }

        [Benchmark]
        public object SpanRef()
        {
            Span<Base> span = array;
            foreach (ref var x in span)
                x = derived;
            return array;
        }
    }
```

|    Method |       Mean |   Error |  StdDev |
|---------- |-----------:|--------:|--------:|
|     Array | 1,490.6 ns | 6.19 ns | 5.17 ns |
| ArrayBase |   600.3 ns | 1.13 ns | 1.06 ns |
|  ArrayInt |   113.1 ns | 0.72 ns | 0.67 ns |
| SpanIndex |   435.9 ns | 1.28 ns | 1.00 ns |
|   SpanRef |   435.1 ns | 0.43 ns | 0.40 ns |

The overhead of array covariance is significant, especially when a derived type instance is used.

I'd suggest to collect the usage data.

**Reply by ufcpp** [reply]
We need [Invariant Arrays](https://github.com/dotnet/runtime/issues/23393)?

**Reply by hez2010** [reply]
> We need [Invariant Arrays](https://github.com/dotnet/runtime/issues/23393)?

We may make an analyzer to analyze popular libraries and see the existing usage of array covariance. If there's evidence showing that the usage is very rare, I think it worth to make a breaking change that remove covariance on array completely.

**Reply by huoyaoyuan** [reply]
> We need [Invariant Arrays](https://github.com/dotnet/runtime/issues/23393)?

It may be less discoverable and require a broad change. Array is a primitive thing and used by everyone. Driving everyone to change to get a performance gain is quite impractical.

**Comment by hez2010** [comment]
I wrote an analyzer to detect array upcast in assemblies, and I ran it against the full .NET + ASP.NET Core assembiles, it turned out that the usage to array upcast is rare, and most of them can be avoided via casting the type of element in array initializer (i.e. change `object[] x = new T[] { new T() ... }` to `object[] x = new object[] { (object)new T()... } `).

The analyzer: https://github.com/hez2010/ArrayUpcastAnalyzer

Here is the analysis report: [report.txt](https://github.com/dotnet/csharplang/files/8212848/report.txt)

/cc: @CyrusNajmabadi

**Reply by jnm2** [reply]
I think making this change would break code I wrote in NUnit.

**Reply by hez2010** [reply]
I did a quick test and found there're only 2 places using array upcast:

```
Analysis for nunit.framework.dll:
Possible array upcast System.Object[] <- System.Type[] at DatapointProvider.GetDataFor: object[] parameters = new Type[0];

Possible array upcast System.Object[] <- System.Collections.IEnumerable[] at UniqueItemsConstraint.TryInferFastPath: object[] parameters = new IEnumerable[1] { actual };
```

To avoid using covariance on array, you can replace them with:
```csharp
object[] parameters = new object[0];
object[] parameters = new object[1] { actual };
```

**Reply by 333fred** [reply]
Try running your analysis on roslyn itself.

**Reply by jnm2** [reply]
I thought there was code that casts `Attribute[]` to `IFoo[]` (via `T[]`), but maybe it was in a compat layer we later removed.

**Reply by hez2010** [reply]
@333fred Minor usage in code analyzer while no usage in csc itself. Here is the analysis result:
```
dotnet run -c Release -- "C:\Program Files\dotnet\sdk\6.0.201\Roslyn\bincore\*.dll"
Analysis for csc.dll:
End of analysis.

Analysis for Microsoft.CodeAnalysis.CSharp.dll:
Possible array upcast System.Object[] <- Microsoft.CodeAnalysis.CSharp.Symbol[] at Binder.CheckViability: object[] args = new Symbol[1] { <>c__DisplayClass660_.unwrappedSymbol };

Possible array upcast System.Object[] <- Microsoft.CodeAnalysis.CSharp.Symbol[] at Binder.CheckViability: object[] args = new Symbol[1] { <>c__DisplayClass660_.unwrappedSymbol };

Possible array upcast System.WeakReference`1[[Microsoft.CodeAnalysis.CSharp.BinderFactory]][] <- System.Object at CSharpCompilation.GetBinderFactory: WeakReference<BinderFactory>[] array = cachedBinderFactories;

Possible array upcast System.Object[] <- System.Xml.Linq.XNode[] at DocumentationCommentCompiler.IncludeElementExpander.Rewrite: object[] content = array2;

Possible array upcast System.Xml.Linq.XNode[] <- System.Xml.Linq.XElement[] at DocumentationCommentCompiler.IncludeElementExpander.RewriteIncludeElement: XNode[] nodes = array;

Possible array upcast Microsoft.CodeAnalysis.CSharp.SingleNamespaceOrTypeDeclaration[] <- Microsoft.CodeAnalysis.CSharp.SingleNamespaceDeclaration[] at DeclarationTreeBuilder.VisitBaseNamespaceDeclaration: SingleNamespaceOrTypeDeclaration[] items = new SingleNamespaceDeclaration[1] { singleNamespaceDeclaration };

Possible array upcast Microsoft.CodeAnalysis.CSharp.BoundExpression[] <- Microsoft.CodeAnalysis.CSharp.BoundLocal[] at LocalRewriter.RewriteMultiDimensionalArrayForEachStatement: BoundExpression[] items = array5;

Possible array upcast Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbol[] <- Microsoft.CodeAnalysis.CSharp.Symbols.NamedTypeSymbol[] at LoweredDynamicOperationFactory.MakeDynamicOperation: Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbol[] typeArguments = new Microsoft.CodeAnalysis.CSharp.Symbols.NamedTypeSymbol[1] { delegateType };

Possible array upcast Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbol[] <- Microsoft.CodeAnalysis.CSharp.Symbols.NamedTypeSymbol[] at LoweredDynamicOperationFactory.DefineCallSiteStorageSymbol: Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbol[] typeArguments = new Microsoft.CodeAnalysis.CSharp.Symbols.NamedTypeSymbol[1] { namedTypeSymbol };

Possible array upcast Microsoft.CodeAnalysis.DiagnosticInfo[] <- Microsoft.CodeAnalysis.CSharp.SyntaxDiagnosticInfo[] at LanguageParser.ParseInterpolatedStringToken: DiagnosticInfo[] diagnostics = new SyntaxDiagnosticInfo[1] { error };

Possible array upcast Microsoft.CodeAnalysis.DiagnosticInfo[] <- Microsoft.CodeAnalysis.CSharp.SyntaxDiagnosticInfo[] at Lexer.AddTrivia: DiagnosticInfo[] errors = GetErrors (0);

Possible array upcast System.Object[] <- System.String[] at Lexer.ScanXmlEntity: object[] array2 = new string[1] { _builder.ToString () };

Possible array upcast System.Object[] <- System.String[] at Lexer.ScanXmlEntity: object[] array2 = new string[1] { c.ToString () };

Possible array upcast System.Object[] <- System.String[] at Lexer.ScanXmlEntity: object[] array2 = new string[1] { c.ToString () };

Possible array upcast System.Byte[] <- System.Object at SourceAttributeData.GetTargetAttributeSignatureIndex: byte[] targetSignature = description.Signatures [i];

End of analysis.

Analysis for Microsoft.CodeAnalysis.dll:
Possible array upcast System.Object[] <- System.Globalization.CultureInfo[] at UICultureUtilities.<>c__DisplayClass3_0.<TryGetCurrentThreadUICultureSetter>b__0: object[] parameters = new CultureInfo[1] { culture };

Possible array upcast System.Object[] <- System.String[] at LocalizableResourceString.GetText: object[] formatArguments = _formatArguments;

Possible array upcast Microsoft.CodeAnalysis.Symbols.ITypeSymbolInternal[] <- `1[] at MetadataDecoder.GetSymbolForILToken: ITypeSymbolInternal[] typeArguments = array;

Possible array upcast System.Byte[] <- System.Object at PEModule.GetTargetAttributeSignatureIndex: byte[] array = description.Signatures [i];

Possible array upcast System.Object[] <- System.String[] at AnalyzerExecutor.CreateAnalyzerExceptionDiagnostic: object[] messageArgs = array;

Possible array upcast System.Object[] <- System.String[] at AnalyzerExecutor.CreateDriverExceptionDiagnostic: object[] messageArgs = array;

Possible array upcast `0[] <- System.Object at SegmentedArray.AlignedSegmentEnumerator.MoveNext: T[] array = _firstSegments [num];

Possible array upcast `0[] <- System.Object at SegmentedArray.AlignedSegmentEnumerator.MoveNext: T[] array2 = _secondSegments [num2];

Possible array upcast `0[] <- System.Object at SegmentedArray.AlignedSegmentEnumerator.MoveNext: T[] array3 = _firstSegments [_completed + _firstOffset >> SegmentedArrayHelper.GetSegmentShift<T> ()];

Possible array upcast `0[] <- System.Object at SegmentedArray.AlignedSegmentEnumerator.MoveNext: T[] array4 = _secondSegments [_completed + _secondOffset >> SegmentedArrayHelper.GetSegmentShift<T> ()];

Possible array upcast `0[] <- System.Object at SegmentedArray.UnalignedSegmentEnumerator.Reverse.MoveNext: T[] array = _firstSegments [num];

Possible array upcast `0[] <- System.Object at SegmentedArray.UnalignedSegmentEnumerator.Reverse.MoveNext: T[] array2 = _secondSegments [num2];

Possible array upcast `0[] <- System.Object at SegmentedArray.UnalignedSegmentEnumerator.MoveNext: T[] array = _firstSegments [num];

Possible array upcast `0[] <- System.Object at SegmentedArray.UnalignedSegmentEnumerator.MoveNext: T[] array2 = _secondSegments [num2];

Possible array upcast `0[] <- System.Object at SegmentedArray.SegmentEnumerator.Reverse.MoveNext: T[] array = _segments [num];

Possible array upcast `0[] <- System.Object at SegmentedArray.SegmentEnumerator.Reverse.MoveNext: T[] array2 = _segments [_completed + _offset >> SegmentedArrayHelper.GetSegmentShift<T> ()];

Possible array upcast `0[] <- System.Object at SegmentedArray.SegmentEnumerator.MoveNext: T[] array = _segments [num];

Possible array upcast `0[] <- System.Object at SegmentedArray.SegmentEnumerator.MoveNext: T[] array2 = _segments [_completed + _offset >> SegmentedArrayHelper.GetSegmentShift<T> ()];

End of analysis.

Analysis for Microsoft.CodeAnalysis.VisualBasic.dll:
Possible array upcast System.Byte[] <- System.Object at SourceAttributeData.GetTargetAttributeSignatureIndex: byte[] array = description.Signatures [i];

End of analysis.

Analysis for vbc.dll:
End of analysis.

Analysis for VBCSCompiler.dll:
End of analysis.
```

**Reply by hez2010** [reply]
> I thought there was code that casts `Attribute[]` to `IFoo[]` (via `T[]`), but maybe it was in a compat layer we later removed.

Rerun it in verbose mode (with code context) to help identify usage:

```
Analysis for nunit.framework.dll:
Possible array upcast System.Object[] <- System.Type[] at DatapointProvider.GetDataFor: object[] parameters = new Type[0];

{
        if (!(methodInfo != null)) {
                continue;
        }
        object obj = (methodInfo.IsStatic ? null : ProviderCache.GetInstanceOf (type2));
        object obj2 = obj;
        object[] parameters = new Type[0];
        foreach (object? item3 in (IEnumerable)methodInfo.Invoke (obj2, parameters)) {
                list.Add (item3);
        }
}

Possible array upcast System.Object[] <- System.Collections.IEnumerable[] at UniqueItemsConstraint.TryInferFastPath: object[] parameters = new IEnumerable[1] { actual };

{
        MethodInfo methodInfo = ItemsCastMethod.MakeGenericMethod (type);
        object[] parameters = new IEnumerable[1] { actual };
        object obj = methodInfo.Invoke (null, parameters);
        if (base.IgnoringCase) {
                if (type == typeof(string)) {
                        return (ICollection)StringsUniqueIgnoringCase ((IEnumerable<string>)obj);
                }
                if (type == typeof(char)) {
                        return (ICollection)CharsUniqueIgnoringCase ((IEnumerable<char>)obj);
                }
        }
        return (ICollection)ItemsUniqueMethod.MakeGenericMethod (type).Invoke (null, new object[1] { obj });
}
```

**Reply by CyrusNajmabadi** [reply]
@hez2010 that does not seem minor at all.

**Reply by hez2010** [reply]
Most usages are `object[] array = new [] { new T()... }`, which can be rewritten to `object[] array = new object[] { new T()... }`. Except this scenario, the left usage is minor.

**Reply by CyrusNajmabadi** [reply]
> the left usage is minor.

I don't see how you can say it's minor.  We're taking advantage of this capability (including in the ways i mention below).  That would mean that removing this capability would be a breaking change immediately with significant consequences.  

It's also unclear what alternative is being proposed for the cases where we are using variance and it's working as intended.  As an example, something basic like this:

```c#
interface IWrapper
{
     ImmutableArray<ISymbol> GetSymbols();
}

class TupleWrapper : IWrapper
{
    private readonly INamedTypeSymbol _tuple;
    public ImmutableArray<ISymbol> GetSymbols()
    {
        ImmutableArray<IFieldSymbol> field = _tuple.TupleFIelds;
        return field.CastUp<ISymbol>();
    }
}
```

Looking in roslyn i counted at least 60+ cases like this.  What would this be replaced with if we do not have variant arrays?

**Reply by hez2010** [reply]
Oops seems that my analyzer doesn't count the `ImmutableArray<T>` case.
We can definitely have a covariant array type, and it's useful, but which really shouldn't be the default behavior of arrays because it will make code much slower even if users don't actually use the covariance feature.

**Reply by CyrusNajmabadi** [reply]
> but which really shouldn't be the default behavior of arrays because it will make code much slower even if users don't actually use the covariance feature.

I disagree.  By default thsi stuff should be convenient and work to make things easy.  If someone needs high perf, they can go to some alternative approach based on data that shows that it is critical.  

We also clearly can't change how arrays work here as that would just break these basic and reasonable cases like the above. 

I would be fine with someone adding some invariant array type for high performance scenarios if called for.  But that would not be in the language, that would be the runtime.  Note: i believe you can simulate this with teh following:

```c#
struct Wrapper<T>
{
    public T Value;
}

Wrapper<object>[] invariant = new Wrapper<object>[10];
invariant[0].Value = "foo";
```

This will have no variance associated with it as this is a struct array.  But it will also have no memory overhead as the struct is the size of the pointer it contains.  Assignments to/from should avoid all variance checks.

**Reply by 333fred** [reply]
> Oops seems that my analyzer doesn't count the ImmutableArray<T> case.

Yeah, this was definitely one of the cases that I was interested in. I know I've written a decent number of locations that use `ImmutableArray.CastUp`. I would also agree with Cyrus that your results (particularly on Roslyn, which is a modern, medium-to-large-sized codebase that I consider well-maintained) indicate that array covariance is used, and the break would be concerning.

**Comment by CyrusNajmabadi** [comment]
Looking in roslyn itself, i see usages of ImmutableArray.CastUp/CastArray, both of which make use of array variance.  So i'll have disagree with both of these points:

1. Convariance on Arrays is useless, and almost no one is actually using it
2. Convariance on Arrays is breaking type system

First, it seems useful for us in several places.  Second, it definitely doesn't have to break the type system.  The way we use it is entirely type-safe.  Without a solution for the ImmutableArray case, i'm not sure we could ever do this.

**Reply by hez2010** [reply]
Besides the type safety issue, there's still a performance issue with covariant array. Covariance makes the runtime need to do extra type check before actually writing to an array, which significantly degrades the performance (see the benchmark above, it's 3x slower compared to invariant span).

**Reply by CyrusNajmabadi** [reply]
Sure.  but i wasn't talking about performance.  Yes, it comes with a perf cost, but as shown above, it is actually useful (And we use it) and it doesn't necessary break the typesystem.  For example, we use it in roslyn without any type-system breakage (despite exposing this through public APIs).  

I get you don't like the perf aspect of it.  That's understandable.  But we can't pretend this isn't both used and baked into core API patterns that .net ships that can be used effectively.

**Comment by bernd5** [comment]
When the concrete array type is known at compile time we could might use that knowlegde in roslyn to generate better code.

I created the following benchmark:
```csharp
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Jobs;
using BenchmarkDotNet.Running;
using System;

class Base
{
}

class Derived : Base
{
}

[SimpleJob(RuntimeMoniker.Net60)]
public class Program
{
    static void Main(string[] args)
    {
        BenchmarkRunner.Run<Program>();
    }

    private Derived derived = new();
    private Base @base = new();

    [Benchmark]
    public object ArrayConstructionSimple()
    {
        var arr = new Base[] {
            derived, derived, derived, derived, derived,
            @base, @base, @base, @base, @base
        };
        
        return arr;
    }

    [Benchmark]
    public object ArrayConstructionWithSpan()
    {
        var arr = new Base[10];
        var span = new Span<Base>(arr);
        span[0] = derived;
        span[1] = derived;
        span[2] = derived;
        span[3] = derived;
        span[4] = derived;
        span[5] = @base;
        span[6] = @base;
        span[7] = @base;
        span[8] = @base;
        span[9] = @base;

        return arr;
    }
}
```

The result is:
```csharp
|                    Method |     Mean |    Error |   StdDev |
|-------------------------- |---------:|---------:|---------:|
|   ArrayConstructionSimple | 55.53 ns | 1.159 ns | 1.586 ns |
| ArrayConstructionWithSpan | 28.69 ns | 0.642 ns | 0.740 ns |
```
